### PR TITLE
Add Deathmatch 3.0

### PR DESCRIPTION
**Pull request by AXDOOMER**
Let's make deathmatches crispier...

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Thank you very much for this pull request!

But, what exactly is "Deathmatch 3.0"? Did you come up yourself with this? Or has this been discussed publically anywhere? Is there some kind of community consent that these rules form "Deathmatch 3.0" and will other ports chime in?

Don't get me wrong, please, but I want to avoid everybody following suit and submit another slightly different set of rules as Deathmatch 4.0, 5.0, 9.0, etc.

**Comment by AXDOOMER** [comment]
I don't know where it first appeared, but I could retrace it in the change log from Doom Legacy 1.27 (dated 1998-12) where it says `deathmatch 3 = deathmatch 2 + weapon stay (dm1+dm2)`.

This set of 'rules' is the default in Zandronum and was also in the now defunct Skulltag, whatever game mode you play (free4all, team deathmatch, capture the flag, etc.). 

Odamex describes it has "newschool deathmatch" (`"Newschool": sv_itemsrespawn 1, sv_weaponstay 1`) on their wiki page about `Sv_gametype`.

People call it "Deathmatch 3.0", since `-altdeath` was "Deathmatch 2.0". People also sometimes call the default deathmatch "Deathmatch 1.0" to refer to the first deathmatch mode explicitly, but I have never seen this in any documentation anywhere. 

It think it would be a nice feature. It's easy to implement (only a few changes required to the source code) and it's the most popular (I can't remember a Zandronum server which didn't use those deathmatch rules). You get the best from both former deathmatch modes: Health pickups and ammo respawns, but weapons stay so when you spawn the weapons will always be there. It think it became popular because on small maps and with >=4 players, if weapons don't stay after pickup then it's unfair for player who just respawned because you have to wait for them to respawn while you're being killed and if items don't respawn, ressources are scarce and  the map becomes boring. 

It's quite useful, for example on E1M1. On deathmatch 2.0, if someone gets the shotgun in the start area or chaingun in at the end of the map and another player respawn in the same area, then this player won't have any weapon close to him, so he will only have his pistol to defend himself. If you don't want this, you have to play in Deathmatch 1.0, but once every item has already been picked up, then you'll eventually die because weapon pickups only give you a small amount of ammo and you will only have enough to kill the other player a few times before you get to the fists. 

Other deathmatch-oriented engines have deathmatch flags (e.g.: barrels respawn, etc.) for full customization and team deathmatch. You even have other game modes such as capture the flag. I believe this is too much. People may ask for other deathmatch modes, but I believe it's fine to ask for deathmatch 3.0 because it's only a small noninvasive change to Crispy's source code that creates an advantageous combinaison of dm1.0 and dm2.0.

**Comment by JNechaevsky** [comment]
I think it's a great feature, a huge step forward for network game.
But will it be safe for some backwards compatibility with Chocolate Doom?

**Comment by AXDOOMER** [comment]
> But will it be safe for some backwards compatibility with Chocolate Doom?

If you select "Deathmatch 3.0" in Crispy and a Chocolate-Doom client tries to start a game with it, then the function ` NET_ValidGameSettings` would probably prevent the game from starting. If you play the other deathmatch modes that both source port have, there won't be any problem. 

It doesn't break demo support. Deathmatch 3.0 demos can be recorded and will playback correctly with compatible source ports.

File: src/doom/p_mobj.c [hunk 1]
```diff
@@ -727,16 +727,16 @@ void P_RespawnSpecials (void)
     int			i;
 
     // only respawn items in deathmatch
-    if (deathmatch != 2)
-	return;	// 
+    if (deathmatch != 2 && deathmatch != 3)
+	return;
```
Comment (fabiangreffrath) [inline src/doom/p_mobj.c]: Please revert the changes to the three `return;` lines. We do not apply cosmetic changes to Choco code in line that we don't change otherwise.

File: src/doom/p_mobj.c [hunk 2]
```diff
@@ -727,16 +727,16 @@ void P_RespawnSpecials (void)
     int			i;
 
     // only respawn items in deathmatch
-    if (deathmatch != 2)
-	return;	// 
+    if (deathmatch != 2 && deathmatch != 3)
```
Comment (fabiangreffrath) [inline src/doom/p_mobj.c]: Please leave a comment here indicating that `deathmatch != 3` is a Crispy-specific change.

File: src/setup/multiplayer.c [hunk 3]
```diff
@@ -266,6 +266,10 @@ static void StartGame(int multiplayer)
         {
             AddCmdLineParameter(exec, "-altdeath");
         }
+        else if (deathmatch == 3)
```
Comment (fabiangreffrath) [inline src/setup/multiplayer.c]: Please leave another comment here indicating that this is a Crispy-specific change.

**Comment by AXDOOMER** [comment]
Changes made.

**Comment by fabiangreffrath** [comment]
Thank you very much!
